### DISCUSSION TITLE
Style - Setter can't read second level property

### DISCUSSION THREAD

**Discussion by stesmi** [AvaloniaUI/Avalonia#6820] [Q&A] [answered]
I've got this simple style:

```
<Style Selector="TextBlock.firstColumn">
    <Setter Property="Width" Value="{Binding #TheLongestFirstColumn.Bounds.Width}"/>
</Style
```
or
```
<Style Selector="TextBlock.firstColumn">
    <Setter Property="Width" Value="{Binding Bounds.Width, ElementName=TheLongestFirstColumn}"/>
</Style
```
Either form works give the same result (and the workaround below can be applied using either form).

One TextBlock is set as:
```
<TextBlock Text="This is the longest line" Name="TheLongestFirstColumn">
```
And in a TextBlock I have:
```
<TextBlock Text="foo" Classes="firstColumn">
```
Expecting the width of that line be the same as the longest line.

Nothing happens, the width is set to 0.

If I instead don't use the setter and apply it directly
```
<TextBlock Text="foo" Width="{Binding #TheLongestFirstColumn.Bounds.Width}">
```
then it works, but obviously I want to use the style.

The workaround is the following:
Let the one that "gives" the setter the value have the following instead:
```
<TextBlock Text="This is the longest line" Name="TheLongestFirstColumn" Tag="{Binding #TheLongestFirstColumn.Bounds.Width}">
```
In essence binding to itself. (I tried using Bounds.Width, RelativeSource={RelativeSource self}, but it didn't work.)
And
```
<Style Selector="TextBlock.firstColumn">
    <Setter Property="Width" Value="{Binding #TheLongestFirstColumn.Tag}"/>
</Style
```

That works.

My thought it that the setter can't bind to a second level property (or I am doing it wrong).

// Stefan

**Comment by MarkusKgit** [comment, marked as answer]
What is the container the items reside in? It works for me just as you posted it in this simple example:

```xaml
<Window ...>
<Window.Styles>
    <Style Selector="TextBlock">
      <Setter Property="Background" Value="Red"/>
      <Setter Property="HorizontalAlignment" Value="Left"/>
      <Setter Property="Margin" Value="5"/>
    </Style>
    <Style Selector="TextBlock.firstColumn">
      <Setter Property="Width" Value="{Binding #TheLongestFirstColumn.Bounds.Width}"/>
    </Style>
  </Window.Styles>
  <StackPanel Orientation="Vertical">
    <TextBlock Text="This is the longest line" Name="TheLongestFirstColumn"/>
    <TextBlock Text="foo" Classes="firstColumn"/>
  </StackPanel>
</Window>
```
With the style:
![Screenshot 2021-10-29 095928](https://user-images.githubusercontent.com/13160892/139398068-5e8d598b-ea5a-41e2-b432-4770695e1408.png)

Without the style:
![Screenshot 2021-10-29 100021](https://user-images.githubusercontent.com/13160892/139398115-f1c6ae7d-3f65-4460-9b04-0859cf8f9a4e.png)

**Comment by stesmi** [comment]
```
                <Grid ColumnDefinitions="Auto, *">
                    <TextBlock Grid.Column="0" Text="Files Processed" Classes="normalLine firstColumn"/>
                    <TextBlock Grid.Column="1" Text="bar"/>
                </Grid>
```
Here is the normalLine setter:
```
        <Style Selector="TextBlock.normalLine">
            <Setter Property="Padding" Value="2"/>
        </Style>
```

**Comment by stesmi** [comment]
Now it gets even weirder.

I uncommented the line that is identical to yours, and commented out my line using Tag and now it works. I spent a few hours on trying to get it to work earlier, but it wouldn't.

I did use Build => Clean Solution inbetween, though. Could that be it? If so, I'm sorry to take up everybody's time.

EDIT: Also tried using ElementName syntax and it works that way as well.

**Reply by MarkusKgit** [reply]
Haha sometimes it just does that, especially with Visual Studio :-) All that matters is that it's working now 👍

**Reply by stesmi** [reply]
Thanx for the help :D

**Reply by MarkusKgit** [reply]
Btw if you want two grid columns to be the same size automatically you can also use `SharedSizeGroup` together with `Grid.IsSharedSizeScope="True"`:
```
<Grid Grid.IsSharedSizeScope="True" >
    <Grid.ColumnDefinitions>
      <ColumnDefinition Width="Auto" SharedSizeGroup="A"/>
      <ColumnDefinition Width="Auto" SharedSizeGroup="A"/>
    </Grid.ColumnDefinitions>
    <TextBlock Grid.Column="0"  Text="foo bar" />
    <TextBlock Grid.Column="1" Text="bar"/>
  </Grid>
```
`HorizontalAlignment` of the `TextBlock`s must be `Strech` in that case

![Screenshot 2021-10-29 104349](https://user-images.githubusercontent.com/13160892/139404886-41c9a0ec-4d00-434f-b94b-d6f8523a7bef.png)
